docs/hpc/08_ood/singularity_with_conda.md

```diff
@@ -0,0 +1,648 @@
+
+#### Use a squashFS Image for Running Jobs
+
+You can use squashFS images similarly to the ext3 images.
+```
+singularity exec --overlay /scratch/<NetID>/pytorch-example/my_pytorch.sqf:ro /scratch/work/public/singularity/cuda12.1.1-cudnn8.9.0-devel-ubuntu22.04.2.sif  /bin/bash -c 'source /ext3/env.sh; python -c "import torch; print(torch.__file__); print(torch.__version__)"'
+
+#example output: /ext3/miniforge3/lib/python3.8/site-packages/torch/__init__.py
+#example output: 1.8.0+cu111
+```
+
+#### Adding Packages to a Full ext3 or squashFS Image 
+
+If the first ext3 overlay image runs out of space or you are using a squashFS conda enviorment, but need to install a new package inside, please copy another writable ext3 overlay image to work together.
+
+Open the first image in read only mode
+```
+cp -rp /scratch/work/public/overlay-fs-ext3/overlay-2GB-100K.ext3.gz .
+gunzip overlay-2GB-100K.ext3.gz
+
+singularity exec --overlay overlay-2GB-100K.ext3 --overlay /scratch/<NetID>/pytorch-example/my_pytorch.ext3:ro /scratch/work/public/singularity/cuda12.1.1-cudnn8.9.0-devel-ubuntu-22.04.2.sif 
+/bin/bash
+source /ext3/env.sh
+pip install tensorboard
+```
+
+***Note:*** Click here for information on how to configure your conda environment.
+
+Please also keep in mind that once the overlay image is opened in default read-write mode, the file will be locked. You will not be able to open it from a new process. Once the overlay is opened either in read-write or read-only mode, it cannot be opened in RW mode from other processes either. For production jobs to run, the overlay image should be open in read-only mode. You can run many jobs at the same time as long as they are run in read-only mode. In this ways, it will protect the computation software environment, software packages are not allowed to change when there are jobs running. 
+
+### Julia Singularity Image
+Singularity can be used to set up a Julia environment.
+
+Create a directory for your julia work, such as `/scratch/<NetID>/julia`, and then change to your home directory. An example is shown below.
+```
+mkdir /home/<NetID>/julia
+cd /home/<NetID>/julia
+```
+
+Copy an overlay image, such as the 2GB 100K overlay, which generally has enough storage for Julia packages. Once copied, unzip to the same folder, rename to julia-pkgs.ext3
+```
+cp -rp /scratch/work/public/overlay-fs-ext3/overlay-2GB-100K.ext3.gz .
+gunzip overlay-2GB-100K.ext3.gz
+mv overlay-2GB-100K.ext3 julia-pkgs.ext3
+```
+
+Copy the following wrapper script in the directory
+```
+cp -rp /share/apps/utils/julia-setup/* .
+```
+
+Now launch writable Singularity overlay to install packages
+```
+module purge
+module load knitro/12.3.0
+module load julia/1.5.3
+
+~/julia/my-julia-writable
+
+using Pkg
+Pkg.add("KNITRO")
+Pkg.add("JuMP")
+```
+
+Now exit from the container to launch a read only version to test (example below)
+```
+~/julia/my-julia
```

Comment:
Any time we're adding a new language that has syntax highlighting support, enable it as described here https://docusaurus.io/docs/markdown-features/code-blocks#supported-languages

Reply:
Cool, will do.  Thanks for the link.  I didn't realize it wasn't supported out of the box.

Reply:
Julia isn't that popular (yet)!